fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java

```diff
@@ -33,22 +36,73 @@
+	 * @param user  The current user of the tool, or null if there isn't any
 	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 * @deprecated Use {@link #getMiningLevel(Tag, BlockState, ItemStack, LivingEntity)} to detect tag and block.
 	 */
-	//TODO: nullable on user once we have an official @Nullable annotation in
-	int getMiningLevel(ItemStack stack, LivingEntity user);
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default int getMiningLevel(ItemStack stack, /* @Nullable */ LivingEntity user) {
+		return 0;
+	}
+
+	/**
+	 * Determines the mining level of the passed stack, which is used for calculating what blocks this tool is allowed to break.
+	 *
+	 * @param tag   The tool tag the item stack is being compared to
+	 * @param state The block to mine
+	 * @param stack The item stack being used to mine the block
+	 * @param user  The current user of the tool, or null if there isn't any
+	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 */
+	// nullable on user once we have an official @Nullable annotation in
+	default int getMiningLevel(Tag<Item> tag, BlockState state, ItemStack stack, /* @Nullable */ LivingEntity user) {
+		return getMiningLevel(stack, user);
+	}
 
 	/**
 	 * Determines the mining speed multiplier of the passed stack, which is one factor in overall mining speed.
 	 *
-	 * @param stack The stack to check on.
-	 * @param user The current user of the tool, or null if there isn't any.
+	 * @param stack The item stack being used to mine the block
+	 * @param user  The current user of the tool, or null if there isn't any
 	 * @return The mining speed multiplier of the item. 8.0 is equal to a diamond pick.
+	 * @deprecated Use {@link #getMiningLevel(Tag, BlockState, ItemStack, LivingEntity)} to detect tag and block.
 	 */
-	//TODO: nullable on user once we have an official @Nullable annotation in
-	float getMiningSpeedMultiplier(ItemStack stack, LivingEntity user);
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default float getMiningSpeedMultiplier(ItemStack stack, /* @Nullable */ LivingEntity user) {
+		return 1.0F;
+	}
+
+	/**
+	 * Determines the mining speed multiplier of the passed stack, which is one factor in overall mining speed.
+	 *
+	 * @param tag   The tool tag the item stack is being compared to
+	 * @param state The block to mine
+	 * @param stack The item stack being used to mine the block
+	 * @param user  The current user of the tool, or null if there isn't any
+	 * @return The mining speed multiplier of the item. 8.0 is equal to a diamond pick.
+	 */
+	// nullable on user once we have an official @Nullable annotation in
+	default float getMiningSpeedMultiplier(Tag<Item> tag, BlockState state, ItemStack stack, /* @Nullable */ LivingEntity user) {
+		return getMiningSpeedMultiplier(stack, user);
+	}
+
+	/**
+	 * Post process the mining speed.
```

Comment:
This needs further explanation with examples where it matters and where it doesn't.